src/main/java/com/atlassian/db/replica/api/DualConnection.java

```diff
@@ -486,6 +520,7 @@ public static Builder builder(
         private CircuitBreaker circuitBreaker = new BreakOnNotSupportedOperations();
         private StateListener stateListener = new NoOpStateListener();
         private Set<String> readOnlyFunctions = new HashSet<>();
+        private boolean compatibleWithPreviousVersion = false;
```

Comment:
So like a "safe mode", which is disabled by default.
So a consumer makes a bump and sets it to `true`. Then they experiment with `false` and if that's good, then they can think about another bump.

So all experiments are coupled together in a given version. And we cannot exclude them from compatibility mode until consumers prove they work. So the release cycle becomes dependent on the consumers.

Not saying that's a blocker.

Reply:
> So the release cycle becomes dependent on the consumers?

If there's a bug in some version, then the customer can keep using the previous version until fixed.

Reply:
Ok, so each release we search and destroy the current `if (compatibleWithPreviousVersion)`?

Reply:
Yes. I think we should add a post-release checklist:
- mark release
- bump the version in the readme
- remove all compatibility checks

Reply:
It forces FFs to be short living. I think it's a good thing.

Reply:
Yep and it doesn't accumulate flags as API, so they can be removed without major release

Reply:
It force limiting WIP bugs, which is good-> LEAN approach, you rollout only 1 change and even if sth is broken you know what changes cause it.

I think about situation when worked 1,2 started in pararel. 1 stream introduced bug, but was first, 2nd stream was slower to implement but it has higher priority.   What's then?

Reply:
Let me try answer myself:

0. we have version 1.0.0
1st changes rolled out with 1.0.1 contain bug.
2nd higher priority changes have to be rolled out as 1.0.2 . 
So maybe `compatibleWithPreviousVersion ` should apply to major/minor changes?
MAJOR.MINOR.PATCH?

Reply:
What's a problem case?